flat_mutation_reader.cc

```diff
@@ -1352,11 +1355,17 @@ flat_mutation_reader downgrade_to_v1(flat_mutation_reader_v2 r) {
         }
         virtual future<> fast_forward_to(const dht::partition_range& pr, db::timeout_clock::time_point timeout) override {
             clear_buffer();
+            // The client didn't receive the paired-up range tombstone change and isn't interested.
+            _rt_assembler.reset();
             _end_of_stream = false;
             return _reader.fast_forward_to(pr, timeout);
         }
         virtual future<> fast_forward_to(position_range pr, db::timeout_clock::time_point timeout) override {
             clear_buffer();
+            if (_rt_assembler.needs_flush()) {
```

Comment:
Judging from the content of the error you're throwing, what you really want to check it is `has_active_tombstone()`, not `needs_flush()`. (the error mentions active range tombstones).

But you're calling `needs_flush()` here because you know the implementation of `needs_flush()`, i.e. you know that `needs_flush() == has_active_tombstone()`.

However, you should not use knowledge about internal implementation of an abstraction when using the abstraction. You should rely only on the interface .

I think it would be better to expose the notion of "active tombstone" to the user of assembler. In other words, make `range_tombstone_assembler::has_active_tombstone()` public and use it here.

OR keep using `needs_flush()`, but modify the content of the message below.

---

It would be also nice to explain here in a comment why is it bad when the client calls `fast_forward_to` in these circumstances?
I'm guessing that this means that the client didn't consume the range till the end before fast_forwarding to another range which is perhaps problematic. Although I'm not sure whether it's really wrong. Does `fast_forward_to` documentation specify that this is not allowed?

Reply:
In documentation it is underspecified:
```
// For any range, all range tombstones relevant for that range which are
// present in the original stream will be emitted. Range tombstones
// emitted before forwarding which overlap with the new range are not
// necessarily re-emitted.
```

If however I don't take this assumption, I don't think it's ever possible to implement `fast_forward_to` for the downgrade_to_v1 correctly, since there are two cases, depending on whether the range tombstones emitted before forwarding which overlap with the new range have been re-emitted.

If there is the re-emission, we need to make sure that the assembler has a reset state. Otherwise it shouldn't.

Reply:
cc @tgrabiec @denesb 

Is it legal for the user of a reader to call `fast_forward_to(pr2)` before fully consuming an earlier fast-forwarded-to range?

Reply:
@tgrabiec @denesb there's one matter which troubles me.
Suppose we have two ranges inside a partition `pr1` and `pr2`, and `start(pr2) > end(pr1)`.
Suppose we have a range tombstone `(x, y)` where `start(pr1) < x < end(pr1)` and `start(pr2) < y < end(pr2)`.
We have a v2 reader.

The v2 flat_mutation_reader documentation states:
```
///
///    4) The sub-stream will not end with a non-neutral active clustered tombstone. All range tombstones are closed.
```
(in the "Intra-partition fast-forwarding mode" section)

Which basically means that the reader will return a `range_tombstone_change` before `end_of_stream` when we read the `pr1` range, closing the `(x, y)` range tombstone.

However, I didn't find anything in the documentation that would state whether it will return another `range_tombstone_change` at the beginning of `pr2` which will re-open the `(x, y)` range tombstone.

I think such a guarantee would be necessary for the users to work correctly, in particular for `downgrade_to_v1`.

Reply:
I think this guarantee already exists, at the beginning of reading a range (even if arrived-to via fast forwarding) we emit all range tombstones the range intersects with, trimmed to said range.

Reply:
Ok, then this guarantee needs to be documented.

In any case, I think the throw in `fast_forward_to` is therefore unnecessary. We should instead reset the assembler.